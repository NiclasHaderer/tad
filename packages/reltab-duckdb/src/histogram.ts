/**
 *
 * Column histograms using reltab-duckdb
 */

import {
  DuckDBDialect,
  NumericSummaryStats,
  QueryExp,
  Schema,
  TableRep,
  cast,
  col,
  constVal,
  divide,
  minus,
  round,
  sqlQuery,
} from "reltab";
import { nice, thresholdSturges } from "./d3utils";

export interface Bin {
  lower: number;
  upper: number;
}

export function binsForColumn(colStats: NumericSummaryStats): number {
  const nullsCount = colStats.pctNull
    ? Math.ceil(colStats.pctNull * colStats.count)
    : 0;
  const valuesCount = colStats.count - nullsCount;
  const numBins = thresholdSturges(valuesCount) + 1;
  return numBins;
}

const doubleType = DuckDBDialect.columnTypes["DOUBLE"];
const intType = DuckDBDialect.columnTypes["INTEGER"];

// Query and metadata needed to form histogram query for a single column
// We return this way so that we can combine multiple histogram queries
// into a single query.
export interface NumericColumnHistogramQuery {
  colId: string;
  histoQuery: QueryExp; // query to compute histogram
  minVal: number;
  maxVal: number;
  niceMinVal: number;
  niceMaxVal: number;
  binCount: number;
  binWidth: number;
}

export function columnHistogramQuery(
  baseQuery: QueryExp,
  querySchema: Schema,
  colId: string,
  colStats: NumericSummaryStats
): NumericColumnHistogramQuery | null {
  const minVal = colStats.min;
  const maxVal = colStats.max;

  if (minVal == null || maxVal == null) {
    return null;
  }
  const binCount = binsForColumn(colStats);

  const [niceMinVal, niceMaxVal] = nice(minVal, maxVal, binCount);

  const binWidth = (niceMaxVal - niceMinVal) / binCount;

  // add a column with bin number:

  const colType = querySchema.columnType(colId);

  const binQuery = baseQuery
    .extend("column", constVal(colId))
    .extend(
      "bin",
      cast(
        round(
          divide(
            minus(
              cast(col(colId), doubleType),
              cast(constVal(niceMinVal), doubleType)
            ),
            cast(constVal(binWidth), doubleType)
          )
        ),
        intType
      )
    );

  const histoQuery = binQuery
    .extend("binCount", constVal(1))
    .groupBy(["column", "bin"], [["count", "binCount"]]);

  const ret = {
    colId,
    histoQuery,
    minVal,
    maxVal,
    niceMinVal,
    niceMaxVal,
    binCount,
    binWidth,
  };
  return ret;
}

export interface NumericColumnHistogramData {
  colId: string;
  niceMinVal: number;
  niceMaxVal: number;
  binCount: number;
  binWidth: number;
  binData: number[];
}

export function getNumericColumnHistogramData(
  colId: string,
  histoQuery: NumericColumnHistogramQuery,
  queryRes: TableRep
): NumericColumnHistogramData {
  const { niceMinVal, niceMaxVal, binCount, binWidth } = histoQuery;
  const numBins = Math.ceil((niceMaxVal - niceMinVal) / binWidth);
  const binData = new Array(numBins).fill(0);
  const { rowData } = queryRes;
  // we could do better by partitioning by column id, but unlikely to be a lot of data for now
  for (const row of rowData) {
    if (row.column === colId) {
      const bin = row.bin as number;
      const binCount = row.binCount;
      binData[bin] = binCount;
    }
  }
  return {
    colId,
    niceMinVal,
    niceMaxVal,
    binCount,
    binWidth,
    binData,
  };
}
