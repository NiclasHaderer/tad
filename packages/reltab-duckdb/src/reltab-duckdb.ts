import { Connection, Database } from "duckdb-async";
import * as log from "loglevel";
import {
  ColumnMetaMap,
  ColumnType,
  DataSourceConnection,
  DataSourceId,
  DataSourceNode,
  DataSourcePath,
  DataSourceProvider,
  DbDataSource,
  DbDriver,
  DuckDBDialect,
  registerProvider,
  Row,
  Schema,
  SQLDialect,
} from "reltab"; // eslint-disable-line
import { initS3 } from "./s3utils";

export * from "./csvimport";

const columnTypes = DuckDBDialect.columnTypes;

let viewCounter = 0;

const genViewName = (): string => `tad_tmpView_${viewCounter++}`;

const typeLookup = (tnm: string): ColumnType => {
  const ret = columnTypes[tnm] as ColumnType | undefined;
  if (ret == null) {
    throw new Error("typeLookup: unknown type name: '" + tnm + "'");
  }
  return ret;
};

/* A little ConnectionPool class because it turns out node-duckdb
 * doesn't allow concurrent queries on a connection.
 */
class ConnectionPool {
  db: Database;
  private pool: Connection[];

  constructor(db: Database) {
    this.db = db;
    this.pool = [];
  }

  async take(): Promise<Connection> {
    if (this.pool.length > 0) {
      return this.pool.pop()!;
    } else {
      const conn = await this.db.connect();
      await initS3(conn);
      return conn;
    }
  }

  giveBack(conn: Connection) {
    this.pool.push(conn);
  }
}

export class DuckDBDriver implements DbDriver {
  readonly displayName: string;
  readonly sourceId: DataSourceId;
  readonly dialect: SQLDialect = DuckDBDialect;
  dbfile: string;
  db: Database;
  connPool: ConnectionPool;

  constructor(dbfile: string, db: Database, dbConn: Connection) {
    this.dbfile = dbfile;
    this.displayName = dbfile;
    this.sourceId = { providerName: "duckdb", resourceId: dbfile };
    this.db = db;
    this.connPool = new ConnectionPool(db);
  }

  async runSqlQuery(query: string): Promise<Row[]> {
    const conn = await this.connPool.take();
    let ret: any;
    try {
      log.info("runSqlQuery:\n", query);
      ret = await conn.all(query);
    } finally {
      this.connPool.giveBack(conn);
    }
    return ret;
  }

  async getDisplayName(): Promise<string> {
    return this.displayName;
  }

  /**
   * Take the rows from a table_info() or DESCRIBE and turn it into
   * a reltab Schema
   * @param metaRows
   */
  schemaFromTableInfo(
    metaRows: Row[],
    columNameKey: string,
    columnTypeKey: string
  ): Schema {
    const extendCMap = (
      columnMetaMap: ColumnMetaMap,
      row: any,
      idx: number
    ): ColumnMetaMap => {
      const displayName = row[columNameKey];
      const columnType = row[columnTypeKey].toLocaleUpperCase();
      const columnMetadata = { displayName, columnType };
      columnMetaMap[displayName] = columnMetadata;
      return columnMetaMap;
    };

    const cmMap = metaRows.reduce(extendCMap, {});
    const columnIds = metaRows.map((r) => r[columNameKey]);
    const schema = new Schema(DuckDBDialect, columnIds as string[], cmMap);
    return schema;
  }

  async getTableSchema(tableName: string): Promise<Schema> {
    const tiQuery = `PRAGMA table_info(${tableName})`;
    const rows = await this.runSqlQuery(tiQuery);

    return this.schemaFromTableInfo(rows, "name", "type");
  }

  async getSqlQuerySchema(sqlQuery: string): Promise<Schema> {
    const describeQuery = `describe ${sqlQuery}`;
    const descRows = await this.runSqlQuery(describeQuery);

    return this.schemaFromTableInfo(descRows, "column_name", "column_type");
  }

  async getRootNode(): Promise<DataSourceNode> {
    const rootNode: DataSourceNode = {
      id: this.dbfile,
      kind: "Database",
      displayName: this.dbfile,
      isContainer: true,
    };
    return rootNode;
  }
  async getChildren(dsPath: DataSourcePath): Promise<DataSourceNode[]> {
    const { path } = dsPath;
    let node: DataSourceNode;
    const tiQuery = `PRAGMA show_tables;`;
    const dbRows = await this.runSqlQuery(tiQuery);
    const tableNames: string[] = dbRows.map((row: Row) => row.name as string);
    const childNodes: DataSourceNode[] = tableNames.map((tableName) => ({
      id: tableName,
      kind: "Table",
      displayName: tableName,
      isContainer: false,
    }));
    return childNodes;
  }

  async getTableName(dsPath: DataSourcePath): Promise<string> {
    const { path } = dsPath;
    if (path.length < 1) {
      throw new Error("getTableName: empty path");
    }
    return path[path.length - 1];
  }
}

const loadExtensions = async (db: Database): Promise<void> => {
  try {
    const ret = await db.exec(`INSTALL 'httpfs'; LOAD 'httpfs'`);
  } catch (err) {
    log.error("caught exception loading extensions: ", err);
    log.error("(ignoring unloadable extensions...)");
  }
};

const duckdbDataSourceProvider: DataSourceProvider = {
  providerName: "duckdb",
  connect: async (resourceId: any): Promise<DataSourceConnection> => {
    const dbfile = resourceId as string;
    const db = await Database.create(dbfile);
    await loadExtensions(db);
    // turn on fast parallel CSV loading:
    // await db.exec("SET experimental_parallel_csv=true;");
    const dbConn = await db.connect();
    const driver = new DuckDBDriver(dbfile, db, dbConn);
    const dsConn = new DbDataSource(driver);
    return dsConn;
  },
};

registerProvider(duckdbDataSourceProvider);
