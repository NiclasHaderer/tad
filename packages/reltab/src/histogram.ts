/**
 *
 * Column histograms using reltab
 */

import { ColumnType, colIsNumeric } from "./ColumnType";
import { DataSourceConnection } from "./DataSource";
import { QueryExp } from "./QueryExp";
import { NumericSummaryStats, Schema } from "./Schema";
import { TableRep } from "./TableRep";
import { nice, thresholdSturges } from "./d3utils";
import { constVal, cast, minus, col, round, divide } from "./defs";
import { DuckDBDialect } from "./dialectRegistry";

export interface Bin {
  lower: number;
  upper: number;
}

export function binsForColumn(colStats: NumericSummaryStats): number {
  const nullsCount = colStats.pctNull
    ? Math.ceil(colStats.pctNull * colStats.count)
    : 0;
  const valuesCount = colStats.count - nullsCount;
  const numBins = thresholdSturges(valuesCount) + 1;
  return numBins;
}

// TODO: adjust to work with any dialect:
// grab dialect.coreColumnTypes.real
const doubleType = DuckDBDialect.columnTypes["DOUBLE"];
const intType = DuckDBDialect.columnTypes["INTEGER"];

// Query and metadata needed to form histogram query for a single column
// We return this way so that we can combine multiple histogram queries
// into a single query.
export interface NumericColumnHistogramQuery {
  colId: string;
  histoQuery: QueryExp; // query to compute histogram
  minVal: number;
  maxVal: number;
  niceMinVal: number;
  niceMaxVal: number;
  binCount: number;
  binWidth: number;
}

export function columnHistogramQuery(
  baseQuery: QueryExp,
  colId: string,
  colType: ColumnType,
  colStats: NumericSummaryStats
): NumericColumnHistogramQuery | null {
  const minVal = colStats.min;
  const maxVal = colStats.max;

  if (minVal == null || maxVal == null) {
    return null;
  }
  const binCount = binsForColumn(colStats);

  const [niceMinVal, niceMaxVal] = nice(minVal, maxVal, binCount);

  const binWidth = (niceMaxVal - niceMinVal) / binCount;

  // add a column with bin number:
  const binQuery = baseQuery
    .extend("column", constVal(colId))
    .extend(
      "bin",
      cast(
        round(
          divide(
            minus(
              cast(col(colId), doubleType),
              cast(constVal(niceMinVal), doubleType)
            ),
            cast(constVal(binWidth), doubleType)
          )
        ),
        intType
      )
    );

  const histoQuery = binQuery
    .extend("binCount", constVal(1))
    .groupBy(["column", "bin"], [["count", "binCount"]]);

  const ret = {
    colId,
    histoQuery,
    minVal,
    maxVal,
    niceMinVal,
    niceMaxVal,
    binCount,
    binWidth,
  };
  return ret;
}

// histogram data for rendering a single column histogram
export interface NumericColumnHistogramData {
  colId: string;
  niceMinVal: number;
  niceMaxVal: number;
  binCount: number;
  binWidth: number;
  binData: number[];
}

/*
 *
 * Given the result of running a histogram query, extract the histogram data for a single column
 */
export function getNumericColumnHistogramData(
  colId: string,
  histoQuery: NumericColumnHistogramQuery,
  queryRes: TableRep
): NumericColumnHistogramData {
  const { niceMinVal, niceMaxVal, binCount, binWidth } = histoQuery;
  const numBins = Math.ceil((niceMaxVal - niceMinVal) / binWidth);
  const binData = new Array(numBins).fill(0);
  const { rowData } = queryRes;
  // we could do better by partitioning by column id, but unlikely to be a lot of data for now
  for (const row of rowData) {
    if (row.column === colId) {
      const bin = row.bin as number;
      const binCount = row.binCount;
      binData[bin] = binCount;
    }
  }
  return {
    colId,
    niceMinVal,
    niceMaxVal,
    binCount,
    binWidth,
    binData,
  };
}

export type ColumnHistogramMap = {
  [colId: string]: NumericColumnHistogramData;
};

export async function getColumnHistogramMap(
  dsConn: DataSourceConnection,
  baseQuery: QueryExp,
  baseSchema: Schema
): Promise<ColumnHistogramMap> {
  const histoMap: ColumnHistogramMap = {};

  // TODO: join this into one mega-query that does a union all
  for (const colId of baseSchema.columns) {
    const colType = baseSchema.columnType(colId);
    if (colIsNumeric(colType)) {
      const colStats = baseSchema.columnStats(colId);
      if (colStats != null) {
        const histoInfo = columnHistogramQuery(
          baseQuery,
          colId,
          colType,
          colStats as NumericSummaryStats
        );
        if (histoInfo) {
          console.log("getting histo for colId: ", colId);
          const histoRes = await dsConn.evalQuery(histoInfo!.histoQuery);
          const histoData = getNumericColumnHistogramData(
            colId,
            histoInfo,
            histoRes
          );
          console.log("histogram data: ", histoData);
          histoMap[colId] = histoData;
        }
      }
    }
  }
  return histoMap;
}
