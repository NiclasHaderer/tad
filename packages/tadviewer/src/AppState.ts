import * as Immutable from "immutable";
import { ViewState } from "./ViewState";
import * as reltab from "reltab";
import { DataSourcePath, DataSourceId } from "reltab";
import { Timer } from "./Timer";
import { Activity } from "./components/defs";
/**
 * Immutable representation of application state
 *
 * Just a single view in a single untabbed window for now.
 */

export interface AppStateProps {
  initialized: boolean; // Has main process initialization completed?

  windowTitle: string; // Usually just the table name or file name

  rtc: reltab.ReltabConnection | null;

  viewState: ViewState | null;
  exportBeginDialogOpen: boolean;
  exportProgressDialogOpen: boolean;
  exportPath: string;
  exportPercent: number;

  viewConfirmDialogOpen: boolean;
  viewConfirmSourcePath: DataSourcePath | null;

  appLoadingTimer: Timer;
  activity: Activity;
  showRecordCount: boolean;
}

const defaultAppStateProps: AppStateProps = {
  initialized: false,
  windowTitle: "",
  rtc: null,
  viewState: null,
  exportBeginDialogOpen: false,
  exportProgressDialogOpen: false,
  exportPath: "",
  exportPercent: 0,
  viewConfirmDialogOpen: false,
  viewConfirmSourcePath: null,
  appLoadingTimer: new Timer(),
  activity: "None",
  showRecordCount: true,
};

export class AppState extends Immutable.Record(defaultAppStateProps) {
  public readonly initialized!: boolean; // Has main process initialization completed?

  public readonly windowTitle!: string; // Usually just the table name or file name

  public readonly rtc!: reltab.ReltabConnection;

  public readonly viewState!: ViewState;
  public readonly exportDialogOpen!: boolean;
  public readonly exportPath!: string;
  public readonly exportPercent!: number;
  public readonly viewConfirmDialogOpen!: boolean;
  public readonly viewConfirmSourcePath!: DataSourcePath | null;
  public readonly appLoadingTimer!: Timer;
  public readonly activity!: Activity;
  public readonly showRecordCount!: boolean;
}
