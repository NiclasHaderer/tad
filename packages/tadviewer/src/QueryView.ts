import * as Immutable from "immutable";
import * as reltab from "reltab";
/*
 * State needed for a scollable view of a reltab query
 */

export interface QueryViewProps {
  query: reltab.QueryExp | null;
  histoMap: reltab.ColumnHistogramMap | null;
  rowCount: number;
  baseRowCount: number;
  filterRowCount: number;
}

const defaultQueryViewProps: QueryViewProps = {
  query: null,
  histoMap: null,
  rowCount: 0,
  // The following fields are just for auxiliary info in footer
  baseRowCount: 0,
  filterRowCount: 0,
};

export class QueryView
  extends Immutable.Record(defaultQueryViewProps)
  implements QueryViewProps
{
  public readonly query!: reltab.QueryExp;
  public readonly histoMap!: reltab.ColumnHistogramMap;
  public readonly rowCount!: number;
  public readonly baseRowCount!: number;
  public readonly filterRowCount!: number;
}
