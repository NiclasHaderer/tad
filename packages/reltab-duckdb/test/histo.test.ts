import * as duckdb from "duckdb-async";
import * as _ from "lodash";
import * as reltab from "reltab";
import {
  asString,
  DataSourceConnection,
  DbDataSource,
  DuckDBDialect,
  tableQuery,
  binsForColumn,
  columnHistogramQuery,
  getNumericColumnHistogramData,
} from "reltab";
import * as reltabDuckDB from "../src/reltab-duckdb";
import * as util from "./testUtils";
import { getFormattedRows } from "./testUtils";

const { col, constVal } = reltab;

const coreTypes = reltab.SQLiteDialect.coreColumnTypes;

let testCtx: DbDataSource;

const q1 = reltab.tableQuery("barttest");

beforeAll(async (): Promise<DataSourceConnection> => {
  const ctx = await reltab.getConnection({
    providerName: "duckdb",
    resourceId: ":memory:",
  });

  testCtx = ctx as DbDataSource;

  const dbds = ctx as DbDataSource;
  const duckDbDriver = dbds.db as reltabDuckDB.DuckDBDriver;

  await reltabDuckDB.nativeCSVImport(
    duckDbDriver.db,
    "test/support/sample.csv"
  );
  await reltabDuckDB.nativeCSVImport(
    duckDbDriver.db,
    "test/support/barttest.csv"
  );

  return testCtx;
});

test("basic column stats and bin count", async () => {
  const qres = await testCtx.evalQuery(q1);

  const tcoeColumnStats = qres.schema.columnMetadata["TCOE"]
    .columnStats as reltab.NumericSummaryStats;

  expect(tcoeColumnStats).toMatchSnapshot();

  const binCount = binsForColumn(tcoeColumnStats);

  console.log("*** number of histogram bins: ", binCount);
});

const intType = DuckDBDialect.columnTypes["INTEGER"];

test("histogram query for column", async () => {
  const schema = await testCtx.getSchema(q1);
  const qres = await testCtx.evalQuery(q1);

  const tcoeColumnStats = qres.schema.columnMetadata["TCOE"]
    .columnStats as reltab.NumericSummaryStats;

  const histoInfo = columnHistogramQuery(q1, "TCOE", intType, tcoeColumnStats);

  console.log("histoInfo: ", histoInfo);

  const histoRes = await testCtx.evalQuery(histoInfo!.histoQuery);
  util.logTable(histoRes);

  const histoData = getNumericColumnHistogramData("TCOE", histoInfo!, histoRes);
  console.log("histogram data:", histoData);
  expect(histoData).toMatchSnapshot();
});
