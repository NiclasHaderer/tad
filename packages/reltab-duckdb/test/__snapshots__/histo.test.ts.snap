// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic column stats and bin count 1`] = `
Object {
  "approxUnique": 23,
  "count": 23,
  "max": 419661,
  "min": 104260,
  "pctNull": 0,
  "statsType": "numeric",
}
`;

exports[`histogram query for column 1`] = `
Object {
  "binCount": 7,
  "binData": Array [
    7,
    3,
    6,
    1,
    4,
    0,
    2,
  ],
  "binWidth": 50000,
  "colId": "TCOE",
  "niceMaxVal": 450000,
  "niceMinVal": 100000,
}
`;
